Single pass ReplaceSource .node performance improvement

brentwilton opened the pull request:
**What kind of change does this PR introduce?**
Performance improvement. Up to 35% faster webpack builds on large solutions.
(small solutions do not spend as much time in ReplaceSource)

**Did you add tests for your changes?**
Added tests to improve coverage around prepending and appending replaces as these were not covered by existing tests.

**If relevant, link to documentation update:**
N/A

**Summary**
Replaces the .node method in ReplaceSource with a version which performs all replacements in a single pass.

**Does this PR introduce a breaking change?**
No

**Other information**
This PR can produce significantly faster build times when processing large files with devtool: 'source-map'

The existing .node method walks the SourceNode tree for every replace each time, twice splitting the SourceNode tree at the replace start and end. In large source maps this can take significant time as the SourceNode constructor will individually .add each child node on each split.

The new method processes all replaces inline, turning on and off an emit flag depending on if you are in the middle of a replace or not.
The PR emits identical module and source map output as webpack 3.4.1 / webpack-sources 1.0.1. (Including use with ModuleConcatenationPlugin which chains ReplaceSource's)
 
For our current Angular prod build using devtool: 'source-map' and 2481 modules, this PR reduces the _chunk asset optimization_ time (average over 3 runs) from ~314 seconds to ~127 seconds, and reduces total webpack build time from ~485 seconds to ~314 seconds. (38.5% or 187 seconds faster)

brentwilton commented:
@sokra @TheLarkInn please let me know if there is anything additional I can provide to help clarify the issue or to explain how the updated ReplaceSource.node() method works.

We are keen to get this PR merged as it makes our CI webpack builds 3 minutes faster.

TheLarkInn commented:
Hey @brentwilton I'm sorry for not seeing this right away. I'm slowly migrating to Octobox for tracking all of these issues and PR's across our orgs' 100+ repos. :-D 

Let me see if I can get this prioritized.

alan-agius4 commented:
Can we have this merged this?

#29

TheLarkInn commented:
This is on my weekend hotlist to merge. Again so sorry for the delay. We want to make sure this doesn't break upstream.

alan-agius4 commented:
Reminder!

brentwilton commented:
@alan-agius4 We have been using a postinstall script in the package.json to manually patch ReplaceSource.js for the last few months while we have been waiting for this PR to be merged.
(I highly recommend implementing this workaround if you have not already. The performance benefits outweigh the hackiness)

With webpack 3.8.1 and the 1.1.0 version of ReplaceSource our CI build takes **47 minutes**. With the PR version, it takes **17 minutes**. (the patch version was temporarily not being applied on our CI build yesterday and everyone thought that the CI server had broken due to builds taking forever)

alan-agius4 commented:
I dont understand and ita frustrating why this PR is not being prioritized! I have the same problem. My CI build iktakes approx 40mins because of this method.

sokra commented:
This is difficult to review, but I take a look soon.

brentwilton commented:
Thanks for the update @sokra. I know you guys have been busy the last few months with progressing webpack 4.

To give a bit more information about the cause of the slowdown and the solution:

The current algorithm does the following:
For _each_ replace we:
1. Traverse the entire source map tree until we find the replace end point
2. Split the source map tree at that point
3. Traverse the entire source map tree until we find the replace start point
4. Split the source map tree at that point
5. Add the new replace text and the end section
6. Move onto the next replacement with the remainder of the tree
		
There are some performance issues with this approach
- We traverse the entire remaining tree twice per replace
- The tree is split twice per replace, at the start and end point of the replace
- When the tree is split, it adds the split sections as children to 2 new SourceNodes
- When you create a new SourceNode, internally it will type check and array.push each child individually to its children array
- On large source maps with a lot of replaces this can take minutes and it thrashes memory while doing so.

The new algorithm changes this to only traverse the source node tree a single time making a copy of it to a new tree.

1. We first sort the list of replacements by start location, and find the first location where we want to replace.
2. When traversing the tree we look for the first replacements start point, and when we hit that point, we turn an emit flag off and stop copying across to the new tree.
3. We add the replacement text as the next node in the copy.
4. We then continue the traversal and look for the replacement end point. On hitting the end point we turn back	on the emit flag and start copying the tree again.
5. When we hit the end point and start emitting again, we then move onto the next replacement in the list and repeat the above until the entire tree is copied.

The ReplacementEnumerator class handles the state of the emit flag and current replacement index.
	
There are some special cases we also deal with:
- Some of the replacements occur outside of the original source map. This is usually webpack adding the imports or require wrappers. We run the _prependStartNodes and _appendRemainingNodes steps to handle this and make the main traversal copy code less complex.
- Replaces can only occur on leaf nodes, and multiple replaces can occur on the same leaf node. Sometimes the start/end points will even overlap so we actually keep recursing into replaceInStringNode after we finish a replace to handle this.
- If you look in the header() method I included some code which tweaks the output to generate identical source maps to the previous version. Identical source map output is good to show we didn't break anything in the change.

We had a situation where a single ReplaceSource.node call was taking 15 minutes on a complex build. The new algorithm reduced that call to 0.12 seconds.

vladimir-tikhonov commented:
Hello, @brentwilton! I've tried use your branch on our project, but unfortunately found a bug. We use `redux-action` package and now webpack produces invalid js code when `devtool` is set to `"source-map"`. Seems like this file is causing issues: https://github.com/reduxactions/redux-actions/blob/master/src/camelCase.js. I've created a repro repository for you here: https://github.com/vladimir-tikhonov/webpack-sources-bug. Let me know if you need something else.

brentwilton commented:
Hi @vladimir-tikhonov. Thanks for trying this PR out and submitting a repro of your issue.
I will pull it down and isolate the cause.

brentwilton commented:
@vladimir-tikhonov  I have pushed a new version. Can you check to see if it resolves your issue ?

The code was double emitting the `);` if it reached the end of the source map while in the middle of a replace.

vladimir-tikhonov commented:
@brentwilton yes, the issue is resolved now. Great job!

alan-agius4 commented:
Can we have this re-reviewed please?

alan-agius4 commented:
Can we have this re-reviewed please?

TheLarkInn commented:
Yes cc @sokra and @ooflorent also.

On Sun, Jan 7, 2018, 12:30 PM Alan Agius <notifications@github.com> wrote:

> Can we have this re-reviewed please?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/webpack/webpack-sources/pull/23#issuecomment-355850264>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADQBMAoKsgJ51a4d4QVMv4ra-hFecZjDks5tISl3gaJpZM4Oe2IV>
> .
>

ooflorent commented on lib/ReplaceSource.js:
```diff
@@ -149,52 +137,208 @@ class ReplaceSource extends Source {
+	}
+
+	_sortReplacementsAscending() {
+		this.replacements.sort(function(a, b) {
```
Could you define the sort function outside of the class?

alan-agius4 replied:
@brentwilton bump

brentwilton replied:
Moved outside of class as requested.

alan-agius4 commented:
Can we merge this obe? Super imporant for me as I have a huge app

TheLarkInn commented:
I think we are waiting on @ooflorent's request, and I'd like @sokra to review as well for this. Given we just released webpack@4.0.0-beta.0 we need to make sure this doesn't cause any breaking changes (even if tests passing).

abhinavsingi commented:
Any idea when shall we be merging this? 
@TheLarkInn @sokra

grahamlucid commented:
Is there any update on getting this issue merged?

Thanks!

grahamlucid commented:
https://media0.giphy.com/media/4SD55a1RnZCdq/giphy.gif

vohzd commented:
Is this getting merged? I mean who doesn't want faster compilation times :D

rpellerin commented:
Would love to see this merged too!

montogeek commented:
@ooflorent @sokra Can we merge this one? It seems it works :)

alan-agius4 commented:
1yr later and this PR is not merged yet! This is just ridiculous.

manniL commented:
Would be awesome to see this merged! :pray:

mlavina commented:
Hey @sokra @ooflorent 

Just checking in as well if this can be merged in and released soon?

sokra commented:
🎉 finally merged

Sorry for the delay, this was difficult to review

sokra commented:
and released as `1.2.0`

alan-agius4 commented:
thank you :)

On Wed, 29 Aug 2018 at 14:10, Tobias Koppers <notifications@github.com>
wrote:

> and released as 1.2.0
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/webpack/webpack-sources/pull/23#issuecomment-416929602>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQv-WteNxcCLROvWIsoYIWXlsaC3wC99ks5uVoShgaJpZM4Oe2IV>
> .
>

eranimo commented:
Is it possible to create a beta release version of webpack with this new version?

evilebottnawi commented:
@eranimo why? It is not major release, just update your lock file. Look on https://github.com/webpack/webpack/blob/master/package.json#L32

edmorley commented:
This decreased the time taken for our warm (babel-loader and uglify-webpack-plugin caches enabled) webpack 4 `--mode production` `devtool: 'source-map'` builds by 30%. Much appreciated - thank you :-)
